### PR TITLE
Null completion field should not throw IAE

### PR DESCRIPTION
**Pull request by tony-dillon**
null completion values should not throw IAE and be added to ignored fields

@jimczi would you mind taking a look at this PR please?  I believe all tests run via `gradlew check` pass now.

many thanks,
Tony

### CODE REVIEW

**Comment by elasticcla** [comment]
Hi @tony-dillon, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch/pull/33268.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by tony-dillon** [comment]
Please note, I have signed the CLA and have received the document via email

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java [hunk 1]
```diff
@@ -435,9 +435,14 @@ public Mapper parse(ParseContext context) throws IOException {
         XContentParser parser = context.parser();
         Token token = parser.currentToken();
         Map<String, CompletionInputMetaData> inputMap = new HashMap<>(1);
+
+        // ignore null values
         if (token == Token.VALUE_NULL) {
-            throw new MapperParsingException("completion field [" + fieldType().name() + "] does not support null values");
-        } else if (token == Token.START_ARRAY) {
+            context.addIgnoredField(fieldType.name());
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java]: I don't think we need to add explicit `null` values to the `_ignored` field. It should be treated as if the field was missing.

File: server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java [hunk 2]
```diff
@@ -448,6 +449,19 @@ public void testFieldValueValidation() throws Exception {
...
+        // null inputs are ignored
+        ParsedDocument nullDoc = defaultMapper.parse(SourceToParse.source("test", "type1", "1", BytesReference
+                .bytes(XContentFactory.jsonBuilder()
+                    .startObject()
+                    .array("completion", Token.VALUE_NULL, Token.VALUE_NULL)
+                    .endObject()),
+            XContentType.JSON));
+        assertThat(doc.docs().size(), equalTo(1));
+        assertNull(doc.docs().get(0).get("completion"));
+        assertNotNull(doc.docs().get(0).getField("_ignored"));
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java]: should be null

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java [hunk 3]
```diff
@@ -435,9 +435,14 @@ public Mapper parse(ParseContext context) throws IOException {
...
+            return null;
+        }
+
+        if (token == Token.START_ARRAY) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java]: can you restore the `} else if` here for readability ?

**Review by jimczi — COMMENTED** [review]
Thanks @tony-dillon , it looks good.
I left a comment regarding the usage of the `_ignored` field.

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java (hunk 1 above)
Comment (tony-dillon) [reply server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java]: stopped adding the field to the ignored list

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java (hunk 3 above)
Comment (tony-dillon) [reply server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java]: reinstated the `else if`

**Comment by tony-dillon** [comment]
Hi @jimczi - thanks for taking the time to review this.

I've addressed your comments, but I'm afraid that there is now an error during an integration test for the `CustomSuggestion` plugin.  

I've attached the report.  The only hint I can see from the report is that there may be a JVM crash:

```[ant:junit4] ERROR: JVM J0 ended with an exception: Forked process returned with error code: 1. Very likely a JVM crash.```

Would you be able to advise how I should tackle this?

[reports.zip](https://github.com/elastic/elasticsearch/files/2336713/reports.zip)

**Review by jimczi — APPROVED** [review]
Thanks @tony-dillon . The jvm crash seems unrelated to this change. I'll trigger a CI build and will merge if it passes.

**Comment by jimczi** [comment]
test this please

**Comment by tony-dillon** [comment]
@jimczi : wow - looks like nothing worked on the `elasticsearch-ci` build.  Almost every single class failed to load.  can't imagine how I've managed that 😄

**Comment by jimczi** [comment]
Can you merge your branch with latest master and push ? I have no idea of what is going on but this is not due to this change.

**Comment by tony-dillon** [comment]
@jimczi merge from master done. thanks again

**Comment by tony-dillon** [comment]
@jimczi 

The build failed again, this time with:

```17:35:43 FAILURE 0.04s J1 | CompletionFieldMapperTests.testFieldValueValidation <<< FAILURES!
17:35:43    > Throwable #1: java.lang.AssertionError: expected null, but was:<stored,indexed,omitNorms,indexOptions=DOCS<_ignored:completion>>```

I wonder, could it be that `null` fields automatically get added to the `ignoredFields` list?

Cheers

File: server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java [hunk 4]
```diff
@@ -448,6 +449,17 @@ public void testFieldValueValidation() throws Exception {
...
         assertThat(ignoredFields.stringValue(), equalTo("completion"));
+
+        // null inputs are ignored
+        ParsedDocument nullDoc = defaultMapper.parse(SourceToParse.source("test", "type1", "1", BytesReference
+                .bytes(XContentFactory.jsonBuilder()
+                    .startObject()
+                    .array("completion", Token.VALUE_NULL, Token.VALUE_NULL)
+                    .endObject()),
+            XContentType.JSON));
+        assertThat(doc.docs().size(), equalTo(1));
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java]: it should be `nullDoc`?

File: server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java [hunk 5]
```diff
@@ -448,6 +449,17 @@ public void testFieldValueValidation() throws Exception {
...
+        assertNull(doc.docs().get(0).get("completion"));
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java]: same here

File: server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java [hunk 6]
```diff
@@ -448,6 +449,17 @@ public void testFieldValueValidation() throws Exception {
...
+        assertNull(doc.docs().get(0).getField("_ignored"));
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java]: and here ;)

**Review by jimczi — COMMENTED** [review]
I left a comment in the tests, they don't use the right variable when asserting.

**Comment by jimczi** [comment]
test this please

**Comment by tony-dillon** [comment]
Ahhhhh schoolboy error!  thank you!

**Comment by tony-dillon** [comment]
Build failed:

```
18:39:12 FAILURE 0.05s J3 | CompletionFieldMapperTests.testFieldValueValidation <<< FAILURES!
18:39:12    > Throwable #1: java.lang.AssertionError: expected null, but was:<VALUE_NULL>
```

should the assert be changed?

```
assertThat(nullDoc.docs().get(0).get("completion"), Token.VALUE_NULL);
```

**Comment by jimczi** [comment]
Yes it should be `assertNull(nullDoc.docs().get(0).get("completion"))`, can you make sure that the test passes locally. If you have troubles to run the tests I can help but you need to run them locally (at least the ones you added) before we run them on CI.

**Comment by tony-dillon** [comment]
Hi @jimczi 

I really appreciate your time with this, and I do always run the tests locally first which actually lead to my last comment.  

I run my local tests like this:

```
/gradlew :server:test -Dtests.class=org.elasticsearch.index.mapper.CompletionFieldMapperTests -Dtests.method="testFieldValueValidation" -Dbuild.snapshot=false -Dlicense.key=./x-pack/qa/smoke-test-plugins-ssl/testnode.pem
```

When I run the tests locally, I get this:

```
  2> REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=8CC2CA7D1B1C6586 -Dtests.class=org.elasticsearch.index.mapper.CompletionFieldMapperTests -Dtests.method="testFieldValueValidation" -Dtests.security.manager=true -Dtests.locale=fr-SC -Dtests.timezone=Africa/Sao_Tome -Dcompiler.java=10 -Druntime.java=10
FAILURE 4.56s | CompletionFieldMapperTests.testFieldValueValidation <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: expected null, but was:<VALUE_NULL>
```
The standout for me is: `expected null, but was:<VALUE_NULL>` which is why I thought maybe I was asserting on the wrong value (i.e. not `assertNull` but `assertThat` and use the `VALUE_NULL` value)

I apologise for taking up your time on this, it's my first contribution and I'm very much just working out how this stuff works 😄 

Many thanks!  
Tony

**Comment by jimczi** [comment]
> I apologise for taking up your time on this, it's my first contribution and I'm very much just working out how this stuff works 😄

No apologies needed, I just wanted to make sure that you can test locally (which is important if you become a contributor ;) ). 

Regarding the failure, we need `assertNull` because we don't add any field in the document if the `completion` is null.

**Comment by tony-dillon** [comment]
if i try it like this:

```
assertThat(nullDoc.docs().get(0).get("completion"), equalTo(Token.VALUE_NULL));
```

I get this:

```
FAILURE 4.44s | CompletionFieldMapperTests.testFieldValueValidation <<< FAILURES!
   > Throwable #1: java.lang.AssertionError:
   > Expected: <VALUE_NULL>
   >      but: was "VALUE_NULL"
```

so I'm not really sure what `null` is in this context.

Thanks again,
T

**Comment by tony-dillon** [comment]
if I change it to:

```
        assertNull(nullDoc.docs().get(0).get("completion"));
```

then I get the error:

```
FAILURE 4.64s | CompletionFieldMapperTests.testFieldValueValidation <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: expected null, but was:<VALUE_NULL>
```

**Comment by jimczi** [comment]
Ok  so the issue is in the json document, instead of:
````
.array("completion", Token.VALUE_NULL, Token.VALUE_NULL)
````
you should do:
````
.nullField("completion")
````
which will create: `"completion" : null`

**Comment by tony-dillon** [comment]
Local tests pass! Thank you Sir!

```
> Task :server:test
==> Test Summary: 1 suite, 1 test
   [junit4] JVM J0:     1.01 ..     9.08 =     8.07s
   [junit4] Execution time total: 9.14 sec.
   [junit4] Tests summary: 1 suite, 1 test

BUILD SUCCESSFUL in 1m 0s
29 actionable tasks: 2 executed, 27 up-to-date
```

File: server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java [hunk 7]
```diff
@@ -440,14 +441,25 @@ public void testFieldValueValidation() throws Exception {
         ParsedDocument doc = defaultMapper.parse(SourceToParse.source("test", "type1", "1", BytesReference
                 .bytes(XContentFactory.jsonBuilder()
                     .startObject()
-                    .array("completion", "   ", "")
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java]: you changed the wrong doc ?

File: server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java [hunk 8]
```diff
@@ -440,14 +441,25 @@ public void testFieldValueValidation() throws Exception {
...
         assertNull(doc.docs().get(0).get("completion"));
         assertNotNull(doc.docs().get(0).getField("_ignored"));
         IndexableField ignoredFields = doc.docs().get(0).getField("_ignored");
         assertThat(ignoredFields.stringValue(), equalTo("completion"));
+
+        // null inputs are ignored
+        ParsedDocument nullDoc = defaultMapper.parse(SourceToParse.source("test", "type1", "1", BytesReference
+                .bytes(XContentFactory.jsonBuilder()
+                    .startObject()
+                    .array("completion", Token.VALUE_NULL, Token.VALUE_NULL)
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java]: it should be ` .nullField("completion")`

**Comment by jimczi** [comment]
@tony-dillon you pushed the change to the wrong doc ;)

**Comment by jimczi** [comment]
test this please

File: server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java [hunk 9]
```diff
@@ -38,6 +38,7 @@
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
 import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.common.xcontent.XContentParser.Token;
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java]: Sorry I missed that we forbid unused import in our build so this needs to be removed.

Comment (tony-dillon) [reply server/src/test/java/org/elasticsearch/index/mapper/CompletionFieldMapperTests.java]: already fixed :) thanks

**Comment by jimczi** [comment]
test this please

**Comment by tony-dillon** [comment]
may I ask @jimczi - what's the recommended way to run the full suite of tests before submitting a PR?

`./gradlew check` or....?

**Comment by jimczi** [comment]
Sure `./gradlew clean check` is what you should run but we can run it on the CI too so you don't need to rerun everything when you fix a small thing in your pr. You can run `./gradlew clean precommit` to check compilation and styles and then run the failing tests locally like you did. It's good to run `./gradlew clean check` before submitting the initial submission though.

**Comment by tony-dillon** [comment]
Thanks for the info.

I notice that you tagged this with with `6.4.1`.  Does that mean that if we finally get this to pass a build, it will make it into the next minor release?

That would be amazing! 😄

**Comment by jimczi** [comment]
test this please

**Comment by jimczi** [comment]
@tony-dillon, I hope you don't mind that I pushed a fix directly in your branch. The change looks good but we need a green build so I triggered a new one.

**Comment by jimczi** [comment]
test this please

**Comment by tony-dillon** [comment]
Of course not @jimczi ! Appreciate you taking the time to dig in and find the issue.

**Comment by tony-dillon** [comment]
Success! @jimczi

**Comment by jimczi** [comment]
I merged in master and 6.x and 6.4.1, thanks @tony-dillon !

**Comment by tony-dillon** [comment]
Thank you @jimczi - you really helped a lot getting this through, and it solves a real-world problem we are having trying to add completion fields to a schema then re-indexing when there are null values present.

Although the code change was small, getting to the point where I could even start to look at the issue was rather slow.  To help others get started developing, I've created a Docker image, and posted an article on Discuss on how to use it.

https://discuss.elastic.co/t/contributing-to-elasticsearch-using-a-docker-image/147162

This should allow anybody to get developing in minutes rather then the hours I spent trying to get the environment setup correctly.

Once again, many thanks for your time and patience getting this live.  
Do you have an ETA for the 6.4.1 release?

Cheers!

**Comment by jimczi** [comment]
You're welcome @tony-dillon.

> Do you have an ETA for the 6.4.1 release?

We don't disclose release dates in advance so the best that I can tell you is that it will probably be soon.

**Pull request merged**